cdc/generation.cc

```diff
@@ -350,34 +350,42 @@ static future<> do_update_streams_description(
     cdc_log.info("CDC description table successfully updated with generation {}.", streams_ts);
 }
 
-void update_streams_description(
+future<> update_streams_description(
         db_clock::time_point streams_ts,
         shared_ptr<db::system_distributed_keyspace> sys_dist_ks,
         noncopyable_function<unsigned()> get_num_token_owners,
         abort_source& abort_src) {
     try {
-        do_update_streams_description(streams_ts, *sys_dist_ks, { get_num_token_owners() }).get();
-    } catch(...) {
+        co_await do_update_streams_description(streams_ts, *sys_dist_ks, { get_num_token_owners() });
+    } catch (...) {
         cdc_log.warn(
             "Could not update CDC description table with generation {}: {}. Will retry in the background.",
             streams_ts, std::current_exception());
 
         // It is safe to discard this future: we keep system distributed keyspace alive.
-        (void)seastar::async([
-            streams_ts, sys_dist_ks, get_num_token_owners = std::move(get_num_token_owners), &abort_src
-        ] {
+        (void)(([
+            _streams_ts = streams_ts,
+            _sys_dist_ks = std::move(sys_dist_ks),
+            _get_num_token_owners = std::move(get_num_token_owners),
+            &_abort_src = abort_src
+        ] () mutable -> future<> {
```

Comment:
Wow... I don't understand what was the motivation here of creating a lambda and then calling it immediately.... Why not just create those variables in the function scope? What does this extra lambda get you?

Reply:
the lambda returns a future which I'm discarding. So the result is that I'm starting a background task

Reply:
I wish we had a nicer idiom to do that...
I'm not sure what to suggest, I'm not experienced enough in the new coroutines stuff :-(

Reply:
I proposed one on slack:
```
template <typename F, typename... Args>
std::invoke_result_t<F, Args...> run_lambda(F&& f, Args... args) {
    auto _f = std::move(f);
    co_return co_await _f(args...);
}

// example usage:
void func(abort_source& abort_src) {
    auto x = std::make_unique<int>(1);
    (void)run_lambda([&abort_src, x = std::move(x)] (int y) -> future<int> {
        co_await sleep_abortable(std::chrono::seconds(1), abort_src);
        co_await sleep_abortable(std::chrono::seconds(1), abort_src);
        co_return *x + y;
    }, 2).then([] (int z) {
        std::cout << z << std::endl;
    });
}
```

however, because `run_lambda` is a coroutine, it needs to allocate its coroutine frame. So we're getting an additional allocation just from calling `run_lambda` (additional to the coroutine frame from `_f`/`f`, so two coroutine frame allocations).

But maybe there's a chance that the `_f(args...)` call will be inlined - in which case we would be dealing with a single coroutine call - or that `run_lambda`'s and `_f`'s coroutine frames will be merged; in both cases we would end up with only one allocation.

Not sure if my hopes for these optimizations even make sense though. Perhaps @avikivity can rebut.